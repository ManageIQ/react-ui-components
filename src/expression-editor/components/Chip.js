import React from 'react';
import PropTypes from 'prop-types';
import { Label } from 'patternfly-react';
import { Dropdown, DropdownToggle, DropdownItem, DropdownSeparator } from '@patternfly/react-core';

export default class Chip extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isFocused: false
    };
  }

  focusedClass = () => (
    (this.props.isFocused && 'focusedChip') || ''
  )

  hideMenu = () => {
    // console.log('ON MENU BLUR');
    this.props.onBlur()
  }

  onBlur = () => {
    // this.setState({isFocused: false});
    setTimeout(() => this.props.onBlur(), 50);
  }

  onFocus = () => {
    this.setState({isFocused: true});
    // this.props.onBlur();
  }

  chipStyle = (type) => {
    switch (type) {
      case "logicalOperator":
        return "primary"
      case "operator":
        return "info"
      case "category":
        return "info"
      case "userinput":
        return "info"
      default:
        return "info"

    }
  }

  setAliasModeWrapper = (bool) => () => {
    this.props.onDoubleClick();
    this.props.setAliasMode(bool);
  }

  setAliasWrapper = (alias) => () => {
    // this.props.onDoubleClick();
    this.props.setAlias(alias);
  }

  render() {
    // console.log('focused: ', this.state.focused)
    // const { a, b, ...rest } = this.props;
    return (
      <Dropdown
        // onToggle={this.onToggle}
        // onSelect={this.hideMenu}
        // onFocus={this.onFocus}
        // onBlur={this.onBlur}
        toggle={
          <span
            ref={this.props.chipRef}
            onKeyDown={this.props.onKeyDown}
            tabIndex="0"
            // onBlur={this.onBlur}
            onFocus={this.props.onFocus}
            className={`${this.focusedClass()}`}
            // {...rest}
          >
            <Label
              // onRemoveClick={this.props.onDelete}
              onClick={this.props.onClick}
              onDoubleClick={this.props.onDoubleClick}
              bsStyle={this.chipStyle(this.props.type)}
            >
              {this.props.label}
            </Label>
          </span>
        }
        // DISEAPEAR BEFORE CLICK
        isOpen={this.props.isFocused}
      >
        <DropdownItem
          component="button"
          // isHovered={this.state.index === i}
          onClick={this.props.onDoubleClick}
        >
          Edit
        </DropdownItem>
        <DropdownItem
          component="button"
          // isHovered={this.state.index === i}
          onClick={this.props.onDelete}
        >
          Delete
        </DropdownItem>
          <DropdownItem
            component="button"
            // isHovered={this.state.index === i}
            onClick={this.setAliasModeWrapper(true)}
          >
            Set Alias
          </DropdownItem> 
          <DropdownItem
            component="button"
            // isHovered={this.state.index === i}
            onClick={this.setAliasWrapper(false)}
          >
            Remove Alias
          </DropdownItem>
      </Dropdown>
    );
  }
}

Chip.propTypes = {
  // onClick: PropTypes.func.isRequired,
  onDoubleClick: PropTypes.func.isRequired,
  onDelete: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  label: PropTypes.string.isRequired,
  onKeyDown: PropTypes.func,
  isFocused: PropTypes.bool.isRequired,
  chipRef: PropTypes.object.isRequired,
};

Chip.defaultProps = {
  onFocus: () => {},
  onBlur: () => {},

};
