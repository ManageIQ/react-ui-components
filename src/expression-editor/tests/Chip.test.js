import React from 'react';
import Chip from '../components/Chip';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';


describe('Chip Component', () => {
  it('has focused class when focused', () => {
    const wrapper = mount(<Chip
      onDelete={jest.fn()}
      onFocus={jest.fn()}
      onDoubleClick={jest.fn()}
      onBlur={jest.fn()}
      onKeyDown={jest.fn()}
      setAliasMode={jest.fn()}
      setAlias={jest.fn()}
      isFocused={true}
      index={0}
      onClick={jest.fn()}
      label='label'
      type='category'
      chipRef={React.createRef()}
    />);

    expect(wrapper.find('span.focusedChip').length > 0).toBeTruthy();
  });

  it('OnClick called when clicked', () => {
    let onClick = jest.fn();
    const wrapper = mount(<Chip
      onDelete={jest.fn()}
      onFocus={jest.fn()}
      onDoubleClick={jest.fn()}
      onBlur={jest.fn()}
      onKeyDown={jest.fn()}
      setAliasMode={jest.fn()}
      setAlias={jest.fn()}
      isFocused={false}
      index={0}
      onClick={onClick}
      label='label'
      type='category'
      chipRef={React.createRef()}
    />);

    wrapper.find('span.pf-c-label').simulate('click');
    expect(onClick).toBeCalled();
  });

  it('OnFocus called when focused', () => {
    let onFocus = jest.fn();
    const wrapper = mount(<Chip
      onDelete={jest.fn()}
      onFocus={onFocus}
      onDoubleClick={jest.fn()}
      onBlur={jest.fn()}
      onKeyDown={jest.fn()}
      setAliasMode={jest.fn()}
      setAlias={jest.fn()}
      isFocused={false}
      index={0}
      onClick={jest.fn()}
      label='label'
      type='category'
      chipRef={React.createRef()}
    />);

    wrapper.find('span.pf-c-label').simulate('focus');
    expect(onFocus).toBeCalled();
  });

  it('onDoubleClick called when double-clicked', () => {
    let onDoubleClick = jest.fn();
    const wrapper = mount(<Chip
      onDelete={jest.fn()}
      onFocus={jest.fn()}
      onDoubleClick={onDoubleClick}
      onBlur={jest.fn()}
      onKeyDown={jest.fn()}
      setAliasMode={jest.fn()}
      setAlias={jest.fn()}
      isFocused={false}
      index={0}
      onClick={jest.fn()}
      label='label'
      type='category'
      chipRef={React.createRef()}
    />);

    wrapper.find('span.pf-c-label').simulate('dblclick');
    expect(onDoubleClick).toBeCalled();
  });

  it('onBlur called when blured', () => {
    let onBlur = jest.fn();
    const wrapper = mount(<Chip
      onDelete={jest.fn()}
      onFocus={jest.fn()}
      onDoubleClick={jest.fn()}
      onBlur={onBlur}
      onKeyDown={jest.fn()}
      setAliasMode={jest.fn()}
      setAlias={jest.fn()}
      isFocused={true}
      index={0}
      onClick={jest.fn()}
      label='label'
      type='category'
      chipRef={React.createRef()}
    />);

    wrapper.find('span.pf-c-label').simulate('blur');
    expect(onBlur).toBeCalled();
  });

  it('onBlur NOT called when clicked on menu item', () => {
    let onBlur = jest.fn();
    const wrapper = mount(<Chip
      onDelete={jest.fn()}
      onFocus={jest.fn()}
      onDoubleClick={jest.fn()}
      onBlur={onBlur}
      onKeyDown={jest.fn()}
      setAliasMode={jest.fn()}
      setAlias={jest.fn()}
      isFocused={true}
      index={0}
      onClick={jest.fn()}
      label='label'
      type='category'
      chipRef={React.createRef()}
    />);

    wrapper.find('button').first().simulate('click');
    expect(onBlur.mock.calls.length).toBe(0);
  });

    it('onDoubleClick called when pressed Enter with default index', () => {
      let onKeyDown = jest.fn();
      let onDoubleClick = jest.fn();
      const wrapper = mount(<Chip
        onDelete={jest.fn()}
        onFocus={jest.fn()}
        onDoubleClick={onDoubleClick}
        onBlur={jest.fn()}
        onKeyDown={onKeyDown}
        setAliasMode={jest.fn()}
        setAlias={jest.fn()}
        isFocused={true}
        index={0}
        onClick={jest.fn()}
        label='label'
        type='category'
        chipRef={React.createRef()}
      />);

      wrapper.find('span.focusedChip').simulate('keydown', {key: 'Enter', keyCode: 13});
      expect(onKeyDown.mock.calls.length).toBe(0);
      expect(onDoubleClick).toBeCalled();
    });

    it('onKeyDown called when other key pressed', () => {
      let onKeyDown = jest.fn();
      let onDoubleClick = jest.fn();
      const wrapper = mount(<Chip
        onDelete={jest.fn()}
        onFocus={jest.fn()}
        onDoubleClick={onDoubleClick}
        onBlur={jest.fn()}
        onKeyDown={onKeyDown}
        setAliasMode={jest.fn()}
        setAlias={jest.fn()}
        isFocused={true}
        index={0}
        onClick={jest.fn()}
        label='label'
        type='category'
        chipRef={React.createRef()}
      />);

      wrapper.find('span.focusedChip').simulate('keydown', {keyCode: 1});
      expect(onKeyDown).toBeCalled();
    });

    it('Navigate and click first item in the menu', () => {
      let onKeyDown = jest.fn();
      let onDoubleClick = jest.fn();
      const wrapper = mount(<Chip
        onDelete={jest.fn()}
        onFocus={jest.fn()}
        onDoubleClick={onDoubleClick}
        onBlur={jest.fn()}
        onKeyDown={onKeyDown}
        setAliasMode={jest.fn()}
        setAlias={jest.fn()}
        isFocused={true}
        index={0}
        onClick={jest.fn()}
        label='label'
        type='category'
        chipRef={React.createRef()}
      />);

      wrapper.find('span.focusedChip').simulate('keydown', {keyCode: 40});
      expect(onKeyDown.mock.calls.length).toBe(0);
      wrapper.find('span.focusedChip').simulate('keydown', {key: 'Enter', keyCode: 13});
      expect(onKeyDown.mock.calls.length).toBe(0);
      expect(onDoubleClick).toBeCalled();
    });

    it('Navigate and click second item in the menu', () => {
      let onKeyDown = jest.fn();
      let onDelete = jest.fn();
      const wrapper = mount(<Chip
        onDelete={onDelete}
        onFocus={jest.fn()}
        onDoubleClick={jest.fn()}
        onBlur={jest.fn()}
        onKeyDown={onKeyDown}
        setAliasMode={jest.fn()}
        setAlias={jest.fn()}
        isFocused={true}
        index={0}
        onClick={jest.fn()}
        label='label'
        type='category'
        chipRef={React.createRef()}
      />);

      wrapper.find('span.focusedChip').simulate('keydown', {keyCode: 40});
      expect(onKeyDown.mock.calls.length).toBe(0);
      wrapper.find('span.focusedChip').simulate('keydown', {keyCode: 40});
      expect(onKeyDown.mock.calls.length).toBe(0);
      wrapper.find('span.focusedChip').simulate('keydown', {key: 'Enter', keyCode: 13});
      expect(onKeyDown.mock.calls.length).toBe(0);
      expect(onDelete).toBeCalled();
    });

    it('Navigate and click third item in the menu', () => {
      let onKeyDown = jest.fn();
      let setAliasMode = jest.fn();
      const wrapper = mount(<Chip
        onDelete={jest.fn()}
        onFocus={jest.fn()}
        onDoubleClick={jest.fn()}
        onBlur={jest.fn()}
        onKeyDown={onKeyDown}
        setAliasMode={setAliasMode}
        setAlias={jest.fn()}
        isFocused={true}
        index={0}
        onClick={jest.fn()}
        label='label'
        type='category'
        chipRef={React.createRef()}
      />);

      wrapper.find('span.focusedChip').simulate('keydown', {keyCode: 40});
      expect(onKeyDown.mock.calls.length).toBe(0);
      wrapper.find('span.focusedChip').simulate('keydown', {keyCode: 40});
      expect(onKeyDown.mock.calls.length).toBe(0);
      wrapper.find('span.focusedChip').simulate('keydown', {keyCode: 40});
      expect(onKeyDown.mock.calls.length).toBe(0);
      wrapper.find('span.focusedChip').simulate('keydown', {key: 'Enter', keyCode: 13});
      expect(onKeyDown.mock.calls.length).toBe(0);
      expect(setAliasMode.mock.calls[0][0]).toBe(true);
      expect(setAliasMode).toBeCalled();
    });

    it('Navigate and click fourth item in the menu', () => {
      let onKeyDown = jest.fn();
      let setAlias = jest.fn();
      const wrapper = mount(<Chip
        onDelete={jest.fn()}
        onFocus={jest.fn()}
        onDoubleClick={jest.fn()}
        onBlur={jest.fn()}
        onKeyDown={onKeyDown}
        setAliasMode={jest.fn()}
        setAlias={setAlias}
        isFocused={true}
        index={0}
        onClick={jest.fn()}
        label='label'
        type='category'
        chipRef={React.createRef()}
      />);

      wrapper.find('span.focusedChip').simulate('keydown', {keyCode: 40});
      expect(onKeyDown.mock.calls.length).toBe(0);
      wrapper.find('span.focusedChip').simulate('keydown', {keyCode: 40});
      expect(onKeyDown.mock.calls.length).toBe(0);
      wrapper.find('span.focusedChip').simulate('keydown', {keyCode: 40});
      expect(onKeyDown.mock.calls.length).toBe(0);
      wrapper.find('span.focusedChip').simulate('keydown', {keyCode: 40});
      expect(onKeyDown.mock.calls.length).toBe(0);
      wrapper.find('span.focusedChip').simulate('keydown', {key: 'Enter', keyCode: 13});
      expect(onKeyDown.mock.calls.length).toBe(0);
      expect(setAlias.mock.calls[0][0]).toBe(false);
      expect(setAlias).toBeCalled();
    });



  it('match snapshot', () => {
    const component = shallow(
      <Chip
        onDelete={jest.fn()}
        onFocus={jest.fn()}
        onDoubleClick={jest.fn()}
        onBlur={jest.fn()}
        onKeyDown={jest.fn()}
        setAliasMode={jest.fn()}
        setAlias={jest.fn()}
        isFocused={true}
        index={0}
        onClick={jest.fn()}
        label='label'
        type='category'
        chipRef={{}}
      />
    );
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
