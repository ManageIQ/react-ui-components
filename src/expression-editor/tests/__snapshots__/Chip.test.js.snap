// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip Component match snapshot 1`] = `
<Dropdown
  className=""
  direction="down"
  dropdownItems={Array []}
  isOpen={true}
  isPlain={false}
  onBlur={[Function]}
  onSelect={[Function]}
  position="left"
  toggle={
    <span
      className="focusedChip"
      onFocus={[MockFunction]}
      onKeyDown={[Function]}
      tabIndex="0"
    >
      <Label
        className=""
        isCompact={false}
        onClick={[MockFunction]}
        onDoubleClick={[MockFunction]}
      >
        label
      </Label>
    </span>
  }
>
  <div
    className="maxDropdownHeght"
  >
    <Item
      className="dropdownFontSize"
      component="button"
      href="#"
      isDisabled={false}
      isHovered={false}
      key="0"
      onClick={[MockFunction]}
      onMouseDown={[Function]}
    >
      Edit
    </Item>
    <Item
      className="dropdownFontSize"
      component="button"
      href="#"
      isDisabled={false}
      isHovered={false}
      key="1"
      onClick={[MockFunction]}
      onMouseDown={[Function]}
    >
      Delete
    </Item>
    <Item
      className="dropdownFontSize"
      component="button"
      href="#"
      isDisabled={false}
      isHovered={false}
      key="2"
      onClick={[Function]}
      onMouseDown={[Function]}
    >
      Set Alias
    </Item>
    <Item
      className="dropdownFontSize"
      component="button"
      href="#"
      isDisabled={false}
      isHovered={false}
      key="3"
      onClick={[Function]}
      onMouseDown={[Function]}
    >
      Remove Alias
    </Item>
  </div>
</Dropdown>
`;
